### PR TITLE
[DI][FrameworkBundle] Hide service ids that start with a dot

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #26630
| License       | MIT
| Doc PR        | -

Now that services are private by default, `debug:container` should display them by default.
In fact, the public/private concept should not trigger a difference in visibility for this command.

Instead, I propose to handle service ids that start with a dot as hidden services.

PR should be ready.

(For reference, I tried using a tag instead in #26891, but that doesn't work in practice when working on the implementation.)

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
ping @symfony/deciders this would be great in 4.1, the current state is clumsy.

**Comment by lyrixx** [comment]
I like the idea. 

But Why did you choose a dot? I think it would be better to choose an underscore even if the meaning of a name starting with a dot in the unix FS is to hide it. 

The underscore is already used in the framework (or other lib) to mean "it's internal / private". I'm thinking to `_locale`, `_format`, `_controller` in the routing Component, `_fragment` in the HttpKernel, `_profiler` in the DebugBundle, `_default` in the DIC, and many more,

**Comment by nicolas-grekas** [comment]
Good question :)
The reason is that an underscored service name cannot be loaded by the YamlFileLoader as they are forbidden. But we have tests that dump + reload configuration to ensure this is allowed. Dumping works, but not loading, for this reason.
So then, if underscore is not allowed, it must be a dot :)

**Comment by sroze** [comment]
I really like the idea as well. And tbh, hiding things that started with `.` is not a new idea so it's a nice choice IMHO.

**Comment by sroze** [comment]
The result of the `debug:container` command is going to be different though, as it will now include the private services. Do we consider this as a BC-break? 🙊

**Comment by nicolas-grekas** [comment]
> Do we consider this as a BC-break?

If we were to answer yes to this question, we would couldn't do any improvement to any command we publish. The answer must be no.

**Comment by lyrixx** [comment]
> So then, if underscore is not allowed, it must be a dot :)

okay, go for it ;)

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php [hunk 1]
```diff
@@ -250,7 +255,7 @@ protected function findDefinitionsByTag(ContainerBuilder $builder, $showPrivate)
             foreach ($builder->findTaggedServiceIds($tag) as $serviceId => $attributes) {
                 $definition = $this->resolveServiceDefinition($builder, $serviceId);
 
-                if (!$definition instanceof Definition || !$showPrivate && !$definition->isPublic()) {
+                if ($showHidden xor '.' === $serviceId[0] ?? null) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Can we use parentheses here? At least I have no idea by reading the code whether `xor` takes precedence over `??` or vice versa.

Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Is this actually what we want at all? Won't we skip non hidden service now if `$showHidden` is `true`?

**Comment by nicolas-grekas** [comment]
> Won't we skip non hidden service now if $showHidden is true?

That's correct: by default, you get non-hidden services, and if you add `--show-hidden`, you get only non-hidden ones. Much more useful IMHO.

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: Can we do that in the other files too that use the same condition? :)

Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: never mind, I was looking at an outdated diff

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php [hunk 2]
```diff
@@ -72,11 +73,6 @@ protected function configure()
 
   <info>php %command.full_name% --types</info>
 
-By default, private services are hidden. You can display all services by
-using the <info>--show-private</info> flag:
-
-  <info>php %command.full_name% --show-private</info>
-
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: You should add some docs about the new flag.

**Review by fabpot — APPROVED** [review]

**Pull request merged**